src/Evolution/DiscontinuousGalerkin/SlopeLimiters/HwenoModifiedSolution.cpp

```diff
@@ -0,0 +1,340 @@
+    noexcept {
+  if (LIKELY(directions_to_exclude.size() == 1)) {
+    // The typical case
+    return a_inverse_matrices.at(primary_direction)
+        .at(directions_to_exclude[0]);
+  } else if (directions_to_exclude.empty()) {
+    // The atypical case is cached in a normally nonsensical slot to reuse the
+    // data structure.
+    return a_inverse_matrices.at(primary_direction).at(primary_direction);
+  } else {
+    ERROR("Asking for an uncached matrix, with multiple excluded neighbors");
+  }
+}
+
+namespace {
+
+template <size_t VolumeDim, size_t DummyIndex>
+const ConstrainedFitCache<VolumeDim>& constrained_fit_cache_impl(
+    const Element<VolumeDim>& element, const Mesh<VolumeDim>& mesh) noexcept {
+  // For the cache to be valid,
+  // - the mesh must always be the same, so we check it below
+  // - the element can be different, as long as it has the same configuration
+  //   of internal/external boundaries. this is handled in the calling code
+  //   because the boundary configuration sets the value of DummyIndex
+  static const Mesh<VolumeDim> mesh_for_cached_matrix = mesh;
+  ASSERT(mesh_for_cached_matrix == mesh,
+         "This call to constrained_fit_cache_impl received a different Mesh\n"
+         "than was previously cached, suggesting that multiple meshes are\n"
+         "used in the computational domain. This is not (yet) supported.\n"
+         "Cached mesh: "
+             << mesh_for_cached_matrix
+             << "\n"
+                "Argument mesh: "
+             << mesh);
+  static const ConstrainedFitCache<VolumeDim> result(element, mesh);
+  return result;
+}
+
+template <size_t VolumeDim, size_t... Is>
+const ConstrainedFitCache<VolumeDim>& constrained_fit_cache(
+    const Element<VolumeDim>& element, const Mesh<VolumeDim>& mesh,
+    std::index_sequence<Is...> /*dummy_indices*/) noexcept {
+  static const std::array<
+      const ConstrainedFitCache<VolumeDim>& (*)(const Element<VolumeDim>&,
+                                                const Mesh<VolumeDim>&),
+      sizeof...(Is)>
+      cache{{&constrained_fit_cache_impl<VolumeDim, Is>...}};
+
+  // Use std::bitset to compute an integer based on the configuration of
+  // internal/external boundaries to the element. This is sort of like a hash
+  // for indexing into the std::array of ConstrainedFitCache's.
+  const size_t index_from_boundary_types = [&element]() noexcept {
+    std::bitset<2 * VolumeDim> bits;
+    for (size_t d = 0; d < VolumeDim; ++d) {
+      for (const Side& side : {Side::Lower, Side::Upper}) {
+        // Index into bitset
+        const size_t bit_index = 2 * d + (side == Side::Lower ? 0 : 1);
+        // Is there a neighbor in this direction?
+        const Direction<VolumeDim> dir(d, side);
+        const bool neighbor_exists =
+            (element.neighbors().find(dir) != element.neighbors().end());
+        bits[bit_index] = neighbor_exists;
+      }
+    }
+    return bits.to_ulong();
```

Comment:
I don't think the standard guarantees that `size_t` is an `unsigned long`. I'm not sure what the right thing to do is, maybe make `index_from_boundary_types` be an `unsigned long`, or I think a `static_cast<size_t>(bits.to_ulong())` would work too (and is probably easier to digest).